File: Python3/0792_number_of_matching_subsequences.py
# https://leetcode.cn/problems/number-of-matching-subsequences/

class Solution:
    def numMatchingSubseq(self, s, words):
        """
        Inputs:
            s: str
            words: list[str]
        Outputs:
            res: int
        """
        idxs = defaultdict(list)
        for i, ch in enumerate(s):
            idxs[ch].append(i)

        res = len(words)
        for word in words:
            cur = -1
            for ch in word:
                # Find a minimum value in idxs_ch that is greater than cur.
                idxs_ch = idxs[ch]
                left, right = 0, len(idxs_ch)-1
                while left <= right:
                    mid = left + (right - left) // 2
                    if idxs_ch[mid] > cur:
                        right = mid - 1
                    else:
                        left = mid + 1
                if left >= len(idxs_ch):
                    res -= 1
                    break

                cur = idxs_ch[left]

        return res

File: Cpp/0792_number_of_matching_subsequences.cpp
// https://leetcode.cn/problems/number-of-matching-subsequences/

class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        vector<vector<int>> idxs(26);
        for (int i = 0; i < s.size(); ++i) {
            idxs[s[i] - 'a'].push_back(i);
        }
        int res = words.size();
        for (string& word: words) {
            // Find a minimum value in idxs[ch-'a'] that is greater than cur.
            int cur = -1;
            for (char ch: word) {
                int n = idxs[ch - 'a'].size();
                int left = 0, right = n - 1;
                while (left <= right) {
                    int mid = left + (right - left) / 2;
                    if (idxs[ch-'a'][mid] > cur) right = mid - 1;
                    else left = mid + 1;
                }
                if (left >= n) {
                    res -= 1;
                    break;
                }
                cur = idxs[ch-'a'][left];
            }
        }
        return res;
    }
};
